x/ecocredit/server/marketplace/buy.go

```diff
@@ -2,10 +2,276 @@ package marketplace
+// - send the coins specified in the bid to the seller.
+func (k Keeper) fillOrder(ctx context.Context, sellOrder *api.SellOrder, buyerAcc sdk.AccAddress, purchaseQty math.Dec, cost sdk.Coin, canPartialFill bool) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	sellOrderQty, err := math.NewDecFromString(sellOrder.Quantity)
+	if err != nil {
+		return err
+	}
+
+	switch sellOrderQty.Cmp(purchaseQty) {
+	case math.LessThan:
+		if !canPartialFill {
+			return ecocredit.ErrInsufficientCredits.Wrapf("cannot purchase %v credits from a sell order that has %s credits", purchaseQty, sellOrder.Quantity)
+		} else {
+			// if we can partially fill, we just delete the sellOrder and take whatever
+			// credits are left from that order.
+			if err := k.stateStore.SellOrderTable().Delete(ctx, sellOrder); err != nil {
+				return err
+			}
+			purchaseQty = sellOrderQty
+		}
+	case math.EqualTo:
+		if err := k.stateStore.SellOrderTable().Delete(ctx, sellOrder); err != nil {
+			return err
+		}
+	case math.GreaterThan:
+		newSellOrderQty, err := sellOrderQty.Sub(purchaseQty)
+		if err != nil {
+			return err
+		}
+		sellOrder.Quantity = newSellOrderQty.String()
+		if err = k.stateStore.SellOrderTable().Update(ctx, sellOrder); err != nil {
+			return err
+		}
+	}
+
+	// remove the credits from the seller's escrowed balance
+	sellerBal, err := k.coreStore.BatchBalanceTable().Get(ctx, sellOrder.Seller, sellOrder.BatchId)
+	if err != nil {
+		return err
+	}
+	escrowBal, err := math.NewDecFromString(sellerBal.Escrowed)
+	if err != nil {
+		return err
+	}
+	escrowBal, err = math.SafeSubBalance(escrowBal, purchaseQty)
+	if err != nil {
+		return err
+	}
+	sellerBal.Escrowed = escrowBal.String()
+	if err = k.coreStore.BatchBalanceTable().Update(ctx, sellerBal); err != nil {
+		return err
+	}
+
+	// update the buyers balance and the batch supply
+	supply, err := k.coreStore.BatchSupplyTable().Get(ctx, sellOrder.BatchId)
+	if err != nil {
+		return err
+	}
+	buyerBal, err := k.coreStore.BatchBalanceTable().Get(ctx, buyerAcc, sellOrder.BatchId)
+	if err != nil {
+		if ormerrors.IsNotFound(err) {
+			buyerBal = &ecocreditv1.BatchBalance{
+				Address:  buyerAcc,
+				BatchId:  sellOrder.BatchId,
+				Tradable: "0",
+				Retired:  "0",
+				Escrowed: "0",
+			}
+		} else {
+			return err
+		}
+	}
+	if sellOrder.DisableAutoRetire { // if auto retire is disabled, we move the credits into the buyer's/supply's tradable balance.
+		tradableBalance, err := math.NewDecFromString(buyerBal.Tradable)
+		if err != nil {
+			return err
+		}
+		tradableBalance, err = math.SafeAddBalance(tradableBalance, purchaseQty)
+		if err != nil {
+			return err
+		}
+		buyerBal.Tradable = tradableBalance.String()
+
+		supplyTradable, err := math.NewDecFromString(supply.TradableAmount)
+		if err != nil {
+			return err
+		}
+		supplyTradable, err = math.SafeAddBalance(supplyTradable, purchaseQty)
+		if err != nil {
+			return err
+		}
+		supply.TradableAmount = supplyTradable.String()
```

Comment:
I think we shouldn't update supply when we escrow/un-escrow credits, just on retirement. We should just have tradable supply = SUM(tradable + escrow balance) I think. Escrowed credits are still tradable

Reply:
ok, would it be cool to wipe this in a follow up PR? this is also done in sell order and test so would add a bit unrelated to the diff

Reply:
https://github.com/regen-network/regen-ledger/issues/959

Reply:
yep 👍